### ISSUE TITLE
repair database --fix-graph throws NPE

### ISSUE THREAD

**Issue by vlasni** [orientechnologies/orientdb#6773]
## OrientDB Version, operating system, or hardware.
- v2.2.10
- v2.2.11
## Operating System

Linux
## Expected behavior and actual behavior

Expected to get an output similar to what's described in http://orientdb.com/docs/last/Console-Command-Repair-Database.html

I am getting just "null pointer exception" with no further details. 

From orientdb.err (partial):

```
2016-10-03 14:41:40:000 INFO  Loading configuration from: /home/develance/orientdb/config/orientdb-server-config.xml... [OServerConfigurationLoaderXml]
2016-10-03 14:41:40:315 INFO  OrientDB Server v2.2.11 (build 2.2.x@r8b3a478e3ca7321a48e7cf0f5991569bbe06ed89; 2016-10-03 09:39:41+0000) is starting up... [OServer]
2016-10-03 14:41:40:322 INFO  Databases directory: /home/develance/orientdb/data [OServer]
2016-10-03 14:41:40:783 INFO  Listening binary connections on 0.0.0.0:2424 (protocol v.36, socket=default) [OServerNetworkListener]
2016-10-03 14:41:40:786 INFO  Listening http connections on 0.0.0.0:2480 (protocol v.10, socket=default) [OServerNetworkListener]
2016-10-03 14:41:40:798 INFO  Installing dynamic plugin 'studio-2.2.zip'... [OServerPluginManager]
2016-10-03 14:41:40:808 INFO  ODefaultPasswordAuthenticator is active [ODefaultPasswordAuthenticator]
2016-10-03 14:41:40:813 INFO  OServerConfigAuthenticator is active [OServerConfigAuthenticator]
2016-10-03 14:41:40:814 INFO  OSystemUserAuthenticator is active [OSystemUserAuthenticator]
2016-10-03 14:41:40:848 INFO  Installed GREMLIN language v.2.6.0 - graph.pool.max=50 [OGraphServerHandler]
2016-10-03 14:41:40:848 INFO  [OVariableParser.resolveVariables] Error on resolving property: distributed [orientechnologies]
2016-10-03 14:41:40:852 WARNI Authenticated clients can execute any kind of code into the server by using the following allowed languages: [sql] [OServerSideScriptInterpreter]
2016-10-03 14:41:40:855 INFO  OrientDB Studio available at http://45.56.122.213:2480/studio/index.html [OServer]
2016-10-03 14:41:40:855 INFO  OrientDB Server is active v2.2.11 (build 2.2.x@r8b3a478e3ca7321a48e7cf0f5991569bbe06ed89; 2016-10-03 09:39:41+0000). [OServer]
2016-10-03 14:43:55:296 WARNI {db=OSystem} Storage 'OSystem' was not closed properly. Will try to recover from write ahead log [OLocalPaginatedStorage]
2016-10-03 14:43:55:297 INFO  {db=OSystem} Looking for last checkpoint... [OLocalPaginatedStorage]
2016-10-03 14:43:55:297 INFO  {db=OSystem} Checkpoints are absent, the restore will start from the beginning. [OLocalPaginatedStorage]
2016-10-03 14:43:55:298 INFO  {db=OSystem} Data restore procedure is started. [OLocalPaginatedStorage]
2016-10-03 14:43:55:305 WARNI {db=OSystem} Non tx operation was used during data modification we will need index rebuild. [OLocalPaginatedStorage]
2016-10-03 14:43:55:311 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OFullCheckpointStartRecord{lsn=OLogSequenceNumber{segment=0, position=5293}, p$
2016-10-03 14:43:55:311 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OCheckpointEndRecord{lsn=OLogSequenceNumber{segment=0, position=5301}} will be$
2016-10-03 14:43:55:318 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OFullCheckpointStartRecord{lsn=OLogSequenceNumber{segment=0, position=14746}, $
2016-10-03 14:43:55:319 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OCheckpointEndRecord{lsn=OLogSequenceNumber{segment=0, position=14770}} will b$
2016-10-03 14:43:55:327 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OFullCheckpointStartRecord{lsn=OLogSequenceNumber{segment=0, position=42277}, $
2016-10-03 14:43:55:328 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OCheckpointEndRecord{lsn=OLogSequenceNumber{segment=0, position=42301}} will b$
2016-10-03 14:43:55:686 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OFullCheckpointStartRecord{lsn=OLogSequenceNumber{segment=0, position=139035},$
2016-10-03 14:43:55:686 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OCheckpointEndRecord{lsn=OLogSequenceNumber{segment=0, position=139059}} will $
2016-10-03 14:43:55:925 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OFullCheckpointStartRecord{lsn=OLogSequenceNumber{segment=0, position=417812},$
2016-10-03 14:43:55:926 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OCheckpointEndRecord{lsn=OLogSequenceNumber{segment=0, position=417836}} will $
2016-10-03 14:43:55:927 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OFullCheckpointStartRecord{lsn=OLogSequenceNumber{segment=0, position=417843},$
2016-10-03 14:43:55:927 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OCheckpointEndRecord{lsn=OLogSequenceNumber{segment=0, position=417867}} will $
2016-10-03 14:43:55:928 WARNI {db=OSystem} Record OFuzzyCheckpointStartRecord{lsn=OLogSequenceNumber{segment=0, position=417874}} com.orientechnologies.orient.core.storage.impl.local.paginated.wa$
2016-10-03 14:43:55:928 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OFuzzyCheckpointEndRecord{lsn=OLogSequenceNumber{segment=0, position=417914}} $
2016-10-03 14:43:55:929 WARNI {db=OSystem} Record OFuzzyCheckpointStartRecord{lsn=OLogSequenceNumber{segment=0, position=417921}} com.orientechnologies.orient.core.storage.impl.local.paginated.wa$
2016-10-03 14:43:55:930 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OFuzzyCheckpointEndRecord{lsn=OLogSequenceNumber{segment=0, position=417961}} $
2016-10-03 14:43:55:930 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OFullCheckpointStartRecord{lsn=OLogSequenceNumber{segment=0, position=417968},$
2016-10-03 14:43:55:931 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OCheckpointEndRecord{lsn=OLogSequenceNumber{segment=0, position=417992}} will $
2016-10-03 14:43:55:932 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OFullCheckpointStartRecord{lsn=OLogSequenceNumber{segment=0, position=417999},$
2016-10-03 14:43:55:934 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OCheckpointEndRecord{lsn=OLogSequenceNumber{segment=0, position=418023}} will $
2016-10-03 14:43:55:935 WARNI {db=OSystem} Record OFuzzyCheckpointStartRecord{lsn=OLogSequenceNumber{segment=0, position=418030}} com.orientechnologies.orient.core.storage.impl.local.paginated.wa$
2016-10-03 14:43:55:936 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OFuzzyCheckpointEndRecord{lsn=OLogSequenceNumber{segment=0, position=418070}} $
2016-10-03 14:43:55:937 WARNI {db=OSystem} Record OFuzzyCheckpointStartRecord{lsn=OLogSequenceNumber{segment=0, position=418077}} com.orientechnologies.orient.core.storage.impl.local.paginated.wa$

```

I tried the repair since whenever I try to access Studio from the browser, I can connect to DB but when selecting "Schema" tab, it reports "Server might be offline".
## Steps to reproduce the problem
- open console
- connect to db (remote)
- run repair database --fix-graph

**Comment by vlasni** [comment]
Just an update: this is production DB and at the moment it's become unusable.

**Comment by vlasni** [comment]
Update: I found in database directory there were old files -I mean indices that were deleted long time ago-. So I moved them out of database directory manually.
Now I am running repair database again and at least now I am getting an output:

`orientdb {db=vlasni2}> repair database --fix-graph
Repair of graph 'remote:api.vlasni.com/vlasni2' is started ...
Scanning 11604293 edges (skipEdges=0)...
+ edges: scanned 100000, removed 0 (estimated remaining time 4941 secs)
+ edges: scanned 200000, removed 0 (estimated remaining time 4732 secs)
+ edges: scanned 300000, removed 0 (estimated remaining time 4634 secs)
`
It's not completed yet, but looks promising.

**Comment by vlasni** [comment]
Update: after edges repair completes, it began vertices repair and throws:
```
Error: java.util.UnknownFormatConversionException: Conversion = ';'

```

**Comment by taburet** [comment]
Fixed the `UnknownFormatConversionException` in 2.2.x and develop branches. You may checkout and build manually or, if you wish, I may create the patched binary distribution for you. There is no need in server upgrade, that was a console only bug.

**Comment by vlasni** [comment]
Hi @taburet, if you could let me know where to download the binary distribution, that would be of great help.
Thanks

**Comment by taburet** [comment]
Here it is https://drive.google.com/file/d/0B9L-XIp8ztKNVDNHZ01hTjR4ajQ/view?usp=sharing

**Comment by vlasni** [comment]
Thank you. Will apply it and let you know if I can get DB back to work.
Just a comment: I found that DB seems to be working fine if accessing it from my code or from console, but when trying to access it from Studio it reports as server is down (which is not).

**Comment by taburet** [comment]
Are there any errors in the log when you try to access it from Studio?

**Comment by taburet** [comment]
Could you please also inspect the log in the JavaScript console of your browser for errors?

**Comment by vlasni** [comment]
Yes, when I try Studio, orientdb.err starts filling with:

```
2016-10-04 07:57:26:512 WARNI {db=OSystem} Database is open by another process [OPaginatedStorageDirtyFlag]
2016-10-04 07:57:26:520 WARNI {db=OSystem} MBean with name com.orientechnologies.orient.core.storage.impl.local.paginated.atomicoperations:id=1603349042,name="OSystem",type=OAtomicOperationsMange$
2016-10-04 07:57:26:521 WARNI {db=OSystem} Storage 'OSystem' was not closed properly. Will try to recover from write ahead log [OLocalPaginatedStorage]
2016-10-04 07:57:26:521 INFO  {db=OSystem} Looking for last checkpoint... [OLocalPaginatedStorage]
2016-10-04 07:57:26:521 INFO  {db=OSystem} Checkpoints are absent, the restore will start from the beginning. [OLocalPaginatedStorage]
2016-10-04 07:57:26:521 INFO  {db=OSystem} Data restore procedure is started. [OLocalPaginatedStorage]
2016-10-04 07:57:26:523 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OFullCheckpointStartRecord{lsn=OLogSequenceNumber{segment=0, position=5293}, p$
2016-10-04 07:57:26:523 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OCheckpointEndRecord{lsn=OLogSequenceNumber{segment=0, position=5301}} will be$
2016-10-04 07:57:26:526 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OFullCheckpointStartRecord{lsn=OLogSequenceNumber{segment=0, position=14746}, $
2016-10-04 07:57:26:526 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OCheckpointEndRecord{lsn=OLogSequenceNumber{segment=0, position=14770}} will b$
2016-10-04 07:57:26:529 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OFullCheckpointStartRecord{lsn=OLogSequenceNumber{segment=0, position=42277}, $
2016-10-04 07:57:26:529 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OCheckpointEndRecord{lsn=OLogSequenceNumber{segment=0, position=42301}} will b$
```

And the browser's console shows:

```
GET http://api.vlasni.com:2480/listDatabases
response:
{
  "errors": [
    {
      "code": 505,
      "reason": 505,
      "content": "java.lang.OutOfMemoryError: Java heap space"
    }
  ]
}

```

**Comment by taburet** [comment]
Did you get to the point when either "Storage data recover was completed" or "Exception during storage data restore" message was written to the database log? Is it possible to share the database?

**Comment by vlasni** [comment]
Will try to share it, compressing it now.

**Comment by laa** [comment]
@vlasni messages which you see in the log means that database was closed incorrectly and try to repair itself. Did you this messages every time when access database ? Did you wait till process of recovery of database content was finished ?

**Comment by vlasni** [comment]
@laa in the past it has happened that db started recovering and it reported in the log when finished. What I've seen this time is that it started the recovery process, it doesn't report as completed but stops logging anything.

**Comment by laa** [comment]
@vlasni what is the last message in the log ? As I understand you stopped the server or there is a chance to get thread dump ?

**Comment by vlasni** [comment]
Below are the last few lines from v2.2.11 log:

```
2016-10-04 08:30:05:930 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OFuzzyCheckpointEndRecord{lsn=OLogSequenceNumber{segment=0, position=425402}} $
2016-10-04 08:30:05:930 WARNI {db=OSystem} Record OFuzzyCheckpointStartRecord{lsn=OLogSequenceNumber{segment=0, position=425409}} com.orientechnologies.orient.core.storage.impl.local.paginated.wa$
2016-10-04 08:30:05:930 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OFuzzyCheckpointEndRecord{lsn=OLogSequenceNumber{segment=0, position=425449}} $
2016-10-04 08:30:05:930 WARNI {db=OSystem} Record OFuzzyCheckpointStartRecord{lsn=OLogSequenceNumber{segment=0, position=425456}} com.orientechnologies.orient.core.storage.impl.local.paginated.wa$
2016-10-04 08:30:05:930 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OFuzzyCheckpointEndRecord{lsn=OLogSequenceNumber{segment=0, position=425496}} $
2016-10-04 08:30:05:930 WARNI {db=OSystem} Record OFuzzyCheckpointStartRecord{lsn=OLogSequenceNumber{segment=0, position=425503}} com.orientechnologies.orient.core.storage.impl.local.paginated.wa$
2016-10-04 08:30:05:930 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OFuzzyCheckpointEndRecord{lsn=OLogSequenceNumber{segment=0, position=425543}} $
2016-10-04 08:41:56:625 WARNI Received signal: SIGTERM [OSignalHandler]
2016-10-04 08:41:56:630 INFO  OrientDB Server is shutting down... [OServer]
```
I have just applied the patch provided by @taburet  and am currently running repair database --fix-graph under v.2.2.12-SNAPSHOT

**Comment by laa** [comment]
@vlasni did server shutdown after you sent kill signal ? There is a risk that you stopped server in the middle of data recovery and as result data will be corrupted.

**Comment by vlasni** [comment]
I didn't kill it directly, I run sudo service orientdb.sh stop

**Comment by vlasni** [comment]
Will upload latest .hprof, it might help I guess

**Comment by laa** [comment]
@vlasni yes I got it server softly stopped but in the middle of data restore. I think we should send a status message every 30 sec. To avoid cases when something is going on but users do not know what it is. not sure that .hprof will help in this concrete case but in general it will be interesting to look at .

**Comment by vlasni** [comment]
I agree about the status message. Sometimes I see Orient consuming resources and nothing being reported.

**Comment by vlasni** [comment]
Here is the link to hprof: https://drive.google.com/file/d/0B1qygo8te6MeOWFMNS0yUU1XQ2M/view?usp=sharing 

And here is the link to DB: https://drive.google.com/file/d/0B1qygo8te6MeU05vd0NwRlhhOVU/view?usp=sharing

**Comment by vlasni** [comment]
Update: repair database --fix-graph completed the edges part with no issues (10M edges). It's currently in the vertices part (2.5M) but it seems to be taking too long, or at least it's not reporting anything on screen.

**Comment by vlasni** [comment]
My apologies for the frequent updates, it is reporting progress now:
```
+ edges: scanned 10700000, removed 0 (estimated remaining time 31 secs)
Scanning edges completed
Scanning 2535330 vertices...
+ vertices: scanned 100000, repaired 0 (estimated remaining time 1003 secs)
+ vertices: scanned 200000, repaired 0 (estimated remaining time 22895 secs)
```

**Comment by taburet** [comment]
Was able to run the provided database:

```
2016-10-05 14:28:09:871 WARNI {db=vlasni2} Storage 'vlasni2' was not closed properly. Will try to recover from write ahead log [OLocalPaginatedStorage]
2016-10-05 14:28:09:871 INFO  {db=vlasni2} Looking for last checkpoint... [OLocalPaginatedStorage]
2016-10-05 14:28:09:872 INFO  {db=vlasni2} FULL checkpoint found. [OLocalPaginatedStorage]
2016-10-05 14:28:09:872 INFO  {db=vlasni2} Data restore procedure from full checkpoint is started. Restore is performed from LSN OLogSequenceNumber{segment=9531, position=74373738} [OLocalPaginatedStorage]
2016-10-05 14:28:09:873 WARNI {db=vlasni2} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OCheckpointEndRecord{lsn=OLogSequenceNumber{segment=9531, position=74373762}} will be skipped during data restore [OLocalPaginatedStorage]
2016-10-05 14:28:09:873 INFO  {db=vlasni2} Storage data recover was completed [OLocalPaginatedStorage]
```

Schema tab is working fine in studio. I suggest you to wait until data recover is finished, "Storage data recover was completed" message will appear in the log. There is chance that the database is broken, since apparently there were multiple interrupted data recover attempts.

**Comment by vlasni** [comment]
@taburet After 24 hours repair database --fix-graph completed, now when I access Studio it starts recovering again, so I will wait till I see the "storage data recover completed" message.

**Comment by vlasni** [comment]
@taburet below is the latest entry in log:

```
2016-10-06 08:17:39:380 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OFuzzyCheckpointEndRecord{lsn=OLogSequenceNumber{segment=0, position=425543}} will be skipped during data restore [OLocalPaginatedStorage]
```
It's been more than an hour since the last entry and still not getting the "storage data recover completed" message. How can I estimate how long would it take to complete?

Thanks

**Comment by taburet** [comment]
Progress is reported every 10000 records processed, there is `storage.wal.reportAfterOperationsDuringRestore` setting which controls that number, but it's too late to tune it, since the recovery is already started, it's better not to interrupt it. In the next yet unreleased version it will be additionally reported every 30 seconds. 

Yes, it's better to stop all other activity on the database to speed things.

**Comment by vlasni** [comment]
ok, I've stopped all applications. 
If I understand a bit of the internals, is it correct to say that if I monitor the .wal file I should see it decrease in size as recovery process moves forward?

**Comment by laa** [comment]
Hi guys,
Data restore is not supposed to be run altogether with data change operations. 
In such case, there is no guarantee that data will be restored correctly.
So in a case of DB crash data, restore process should be finished and only after that all operations should be started. In the current situation, I do not think that data will be restored correctly. 
Anyway could you send us thread dump to get what is going on inside of DB ?

**Comment by vlasni** [comment]
Here is a link to a dump I just created:  https://drive.google.com/file/d/0B1qygo8te6MeT1VvX3NsZnJEXzQ/view?usp=sharing

**Comment by laa** [comment]
@vlasni seems like restore is finished, dunno why message about the end of restore procedure was not printed. Could you do following then? Take a copy of the database which you sent to open it @taburet, wait till restore is finished and then run repair database command?

**Comment by vlasni** [comment]
Here is the link to db: https://drive.google.com/file/d/0B1qygo8te6MeaXVtcmNKNkk2bUU/view?usp=sharing 
I stopped Orient before gzipping db, deleted log files, restarted Orient. Log looks OK:

```
Starting OrientDB server daemon...
develance@vlasniops01:~/orientdb/orientdb-community-2.2.12-SNAPSHOT/log$ tail -f orientdb.err 
2016-10-06 10:38:52:249 INFO  Listening http connections on 0.0.0.0:2480 (protocol v.10, socket=default) [OServerNetworkListener]
2016-10-06 10:38:52:262 INFO  Installing dynamic plugin 'studio-2.2.zip'... [OServerPluginManager]
2016-10-06 10:38:52:274 INFO  ODefaultPasswordAuthenticator is active [ODefaultPasswordAuthenticator]
2016-10-06 10:38:52:275 INFO  OServerConfigAuthenticator is active [OServerConfigAuthenticator]
2016-10-06 10:38:52:276 INFO  OSystemUserAuthenticator is active [OSystemUserAuthenticator]
2016-10-06 10:38:52:300 INFO  Installed GREMLIN language v.2.6.0 - graph.pool.max=50 [OGraphServerHandler]
2016-10-06 10:38:52:300 INFO  [OVariableParser.resolveVariables] Error on resolving property: distributed [orientechnologies]
2016-10-06 10:38:52:305 WARNI Authenticated clients can execute any kind of code into the server by using the following allowed languages: [sql] [OServerSideScriptInterpreter]
2016-10-06 10:38:52:309 INFO  OrientDB Studio available at http://45.56.122.213:2480/studio/index.html [OServer]
2016-10-06 10:38:52:309 INFO  OrientDB Server is active v2.2.12-SNAPSHOT (build 2.2.x@rdfed502689481d7ec48b548e54997faf319788d5; 2016-10-04 10:59:10+0000). [OServer]
```
Now I try to access if from Studio, and the same behavior as before. This is the beginning of the log:

```

2016-10-06 10:41:56:738 WARNI ->190.246.133.85: Command not found: GET.studio/scripts/filesize.min.js.map [ONetworkProtocolHttpDb]
2016-10-06 10:42:07:099 WARNI {db=OSystem} Storage 'OSystem' was not closed properly. Will try to recover from write ahead log [OLocalPaginatedStorage]
2016-10-06 10:42:07:100 INFO  {db=OSystem} Looking for last checkpoint... [OLocalPaginatedStorage]
2016-10-06 10:42:07:100 INFO  {db=OSystem} Checkpoints are absent, the restore will start from the beginning. [OLocalPaginatedStorage]
2016-10-06 10:42:07:101 INFO  {db=OSystem} Data restore procedure is started. [OLocalPaginatedStorage]
2016-10-06 10:42:07:207 WARNI {db=OSystem} Non tx operation was used during data modification we will need index rebuild. [OLocalPaginatedStorage]
2016-10-06 10:42:07:217 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OFullCheckpointStartRecord{lsn=OLogSequenceNumber{segment=0, position=5293}, previousCheckpoint=null} will be skipped during data restore [OLocalPaginatedStorage]
2016-10-06 10:42:07:218 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OCheckpointEndRecord{lsn=OLogSequenceNumber{segment=0, position=5301}} will be skipped during data restore [OLocalPaginatedStorage]
2016-10-06 10:42:07:225 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OFullCheckpointStartRecord{lsn=OLogSequenceNumber{segment=0, position=14746}, previousCheckpoint=OLogSequenceNumber{segment=0, position=5293}} will be skipped during data restore [OLocalPaginatedStorage]
2016-10-06 10:42:07:226 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OCheckpointEndRecord{lsn=OLogSequenceNumber{segment=0, position=14770}} will be skipped during data restore [OLocalPaginatedStorage]
2016-10-06 10:42:07:436 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OFullCheckpointStartRecord{lsn=OLogSequenceNumber{segment=0, position=42277}, previousCheckpoint=OLogSequenceNumber{segment=0, position=14746}} will be skipped during data restore [OLocalPaginatedStorage]
2016-10-06 10:42:07:437 WARNI {db=OSystem} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OCheckpointEndRecord{lsn=OLogSequenceNumber{segment=0, position=42301}} will be skipped during data restore [OLocalPaginatedStorage]
```

**Comment by vlasni** [comment]
I've also uploaded OSystem db as it seems to be the problem: https://drive.google.com/file/d/0B1qygo8te6MeWUlFTXItc0dZRTA/view?usp=sharing

**Comment by vlasni** [comment]
As I see OSystem is being reported in the log, I tried to connect to it from console, and see what happened:
```
orientdb> connect remote:api.vlasni.com/OSystem root myPasswd

Disconnecting from the database [null]...OK
Connecting to database [remote:api.vlasni.com/OSystem] with user 'root'...
Error: com.orientechnologies.orient.enterprise.channel.binary.OResponseProcessingException: Exception during response processing

Error: java.lang.OutOfMemoryError: Java heap space
```

**Comment by vlasni** [comment]
Update: I removed from data/OSystem some files I understood were not needed (based on their date):

- OFunction.name.hnb
- ORole.name.nbt
- OUser.name.nbt
- command-cache.json
- database.ocf

Now when accessing from Studio I get the following in the log, but Studio opens and allows me to see the schema:
```
2016-10-06 11:44:24:650 WARNI ->190.246.133.85: Command not found: GET.studio/scripts/filesize.min.js.map [ONetworkProtocolHttpDb]
2016-10-06 11:44:35:467 SEVER isAuthorized() Exception: Cannot open local storage '/home/develance/orientdb/data/OSystem' with mode=rw
	DB name="OSystem" [OSystemUserAuthenticator]
```
If I try to connect from console:
```
orientdb> connect remote:api.vlasni.com/OSystem root myPasswd

Connecting to database [remote:api.vlasni.com/OSystem] with user 'root'...
Error: com.orientechnologies.orient.core.exception.OStorageException: Cannot open local storage '/home/develance/orientdb/data/OSystem' with mode=rw
	DB name="OSystem"
	DB name="OSystem"

Error: com.orientechnologies.orient.core.exception.OStorageException: Storage OSystem is not opened.
	DB name="OSystem"

```

**Comment by laa** [comment]
@vlasni remove OSystem folder completely, and the database will be automatically recreated it contains only service information.

**Comment by vlasni** [comment]
@laa that worked perfect. I can now access Studio with no issues nor anything unusual in the logs.
Thank you very much!

**Comment by taburet** [comment]
Closing the issue, feel free to reopen, if the problem reappears.

**Comment by vlasni** [comment]
@taburet  I am again experiencing issues with DB. I did as before: removed OSystem contents but this time when trying to connect to OSystem everything works fine, but when attempting to connect to db vlasni2, I can't:
```
orientdb> connect remote:api.vlasni.com/vlasni2 root myPasswd

Connecting to database [remote:api.vlasni.com/vlasni2] with user 'root'...
Error: com.orientechnologies.orient.enterprise.channel.binary.OResponseProcessingException: Exception during response processing

Error: java.lang.OutOfMemoryError: Java heap space
```

This is what log shows:
```
2016-10-08 14:54:41:588 WARNI {db=vlasni2} Storage 'vlasni2' was not closed properly. Will try to recover from write ahead log [OLocalPaginatedStorage]
2016-10-08 14:54:41:589 INFO  {db=vlasni2} Looking for last checkpoint... [OLocalPaginatedStorage]
2016-10-08 14:54:41:590 INFO  {db=vlasni2} Found FUZZY checkpoint. [OLocalPaginatedStorage]
2016-10-08 14:54:41:590 INFO  {db=vlasni2} Data restore procedure from FUZZY checkpoint is started. [OLocalPaginatedStorage]
2016-10-08 14:54:41:591 WARNI {db=vlasni2} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OFuzzyCheckpointEndRecord{lsn=OLogSequenceNumber{segment=10509, position=57367041}} will be skipped during data restore [OLocalPaginatedStorage]
2016-10-08 15:00:49:831 WARNI {db=vlasni2} Database is open by another process [OPaginatedStorageDirtyFlag]
2016-10-08 15:00:49:841 WARNI {db=vlasni2} MBean with name com.orientechnologies.orient.core.storage.impl.local.paginated.atomicoperations:id=950537827,name="vlasni2",type=OAtomicOperationsMangerMXBean has already registered. Probably your system was not shutdown correctly or you have several running applications which use OrientDB engine inside [OAtomicOperationsManager]
2016-10-08 15:00:49:841 WARNI {db=vlasni2} Storage 'vlasni2' was not closed properly. Will try to recover from write ahead log [OLocalPaginatedStorage]
2016-10-08 15:00:49:841 INFO  {db=vlasni2} Looking for last checkpoint... [OLocalPaginatedStorage]
2016-10-08 15:00:49:842 INFO  {db=vlasni2} Found FUZZY checkpoint. [OLocalPaginatedStorage]
2016-10-08 15:00:49:842 INFO  {db=vlasni2} Data restore procedure from FUZZY checkpoint is started. [OLocalPaginatedStorage]
2016-10-08 15:00:49:842 WARNI {db=vlasni2} Record com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OFuzzyCheckpointEndRecord{lsn=OLogSequenceNumber{segment=10509, position=57367041}} will be skipped during data restore [OLocalPaginatedStorage]
```
I am not getting any "storage data recover completed" message in the log (I kept it in that status 24 hours and nothing happened).
I really need to get this db back, any suggestions please?

**Comment by vlasni** [comment]
Update: I have db back working. Not sure why it worked, but this is what I did:

- created new db from Studio
- shutdown server
- copied all contents of old db directory to new db directory
- started server

Now I am using new db. No recovery messages shown in log, everything works fine.

**Comment by vlasni** [comment]
Hi guys, I am back again having issues with my DB.

```
2016-10-11 09:52:49:152 INFO  Listening http connections on 0.0.0.0:2480 (protocol v.10, socket=default) [OServerNetworkListener]
2016-10-11 09:52:49:170 INFO  Installing dynamic plugin 'studio-2.2.zip'... [OServerPluginManager]
2016-10-11 09:52:49:175 INFO  ODefaultPasswordAuthenticator is active [ODefaultPasswordAuthenticator]
2016-10-11 09:52:49:175 INFO  OServerConfigAuthenticator is active [OServerConfigAuthenticator]
2016-10-11 09:52:49:176 INFO  OSystemUserAuthenticator is active [OSystemUserAuthenticator]
2016-10-11 09:52:49:201 INFO  Installed GREMLIN language v.2.6.0 - graph.pool.max=50 [OGraphServerHandler]
2016-10-11 09:52:49:202 INFO  [OVariableParser.resolveVariables] Error on resolving property: distributed [orientechnologies]
2016-10-11 09:52:49:205 WARNI Authenticated clients can execute any kind of code into the server by using the following allowed languages: [sql] [OServerSideScriptInterpreter]
2016-10-11 09:52:49:208 INFO  OrientDB Studio available at http://45.56.122.213:2480/studio/index.html [OServer]
2016-10-11 09:52:49:208 INFO  OrientDB Server is active v2.2.12-SNAPSHOT (build 2.2.x@rdfed502689481d7ec48b548e54997faf319788d5; 2016-10-04 10:59:10+0000). [OServer]
2016-10-11 09:55:28:804 WARNI {db=vlasni3} Storage 'vlasni3' was not closed properly. Will try to recover from write ahead log [OLocalPaginatedStorage]
2016-10-11 09:55:28:804 INFO  {db=vlasni3} Looking for last checkpoint... [OLocalPaginatedStorage]
```
It's been 42 minutes so far since last output in log and DB is inaccessible at the moment. Trying to connect from console throws "Error: java.lang.OutOfMemoryError: Java heap space".

Background: application logs started to show entries like this one:

```

2016-10-11 09:12:42,853 TRACE Trying TCCL for class com.orientechnologies.orient.client.remote.ORemoteConnectionPool.
2016-10-11 09:12:42.852 [http-nio-9090-exec-38] ERROR com.develance.vlastni.api.auth.AuthServlet 60 doGet - com.orientechnologies.common.io.OIOException: Error on connecting to api.vlasni.com:242$
        at com.orientechnologies.orient.client.remote.ORemoteConnectionPool.createNetworkConnection(ORemoteConnectionPool.java:68) ~[orientdb-client-2.2.11.jar:2.2.11]
        at com.orientechnologies.orient.client.remote.ORemoteConnectionPool.createNewResource(ORemoteConnectionPool.java:74) ~[orientdb-client-2.2.11.jar:2.2.11]
        at com.orientechnologies.orient.client.remote.ORemoteConnectionPool.createNewResource(ORemoteConnectionPool.java:19) ~[orientdb-client-2.2.11.jar:2.2.11]
        at com.orientechnologies.common.concur.resource.OResourcePool.getResource(OResourcePool.java:96) ~[orientdb-core-2.2.11.jar:2.2.11]
        at com.orientechnologies.orient.client.remote.ORemoteConnectionPool.acquire(ORemoteConnectionPool.java:98) ~[orientdb-client-2.2.11.jar:2.2.11]
        at com.orientechnologies.orient.client.remote.ORemoteConnectionManager.acquire(ORemoteConnectionManager.java:101) ~[orientdb-client-2.2.11.jar:2.2.11]
        at com.orientechnologies.orient.client.remote.OStorageRemote.getNetwork(OStorageRemote.java:2113) ~[orientdb-client-2.2.11.jar:2.2.11]
        at com.orientechnologies.orient.client.remote.OStorageRemote.baseNetworkOperation(OStorageRemote.java:211) ~[orientdb-client-2.2.11.jar:2.2.11]
        at com.orientechnologies.orient.client.remote.OStorageRemote.networkOperation(OStorageRemote.java:193) ~[orientdb-client-2.2.11.jar:2.2.11]
        at com.orientechnologies.orient.client.remote.OStorageRemote.command(OStorageRemote.java:1129) ~[orientdb-client-2.2.11.jar:2.2.11]
        at com.orientechnologies.orient.core.command.OCommandRequestTextAbstract.execute(OCommandRequestTextAbstract.java:69) ~[orientdb-core-2.2.11.jar:2.2.11]
        at com.tinkerpop.blueprints.impls.orient.OrientGraphCommand.execute(OrientGraphCommand.java:49) ~[orientdb-graphdb-2.2.11.jar:2.2.11]
        at com.develance.vlastni.api.model.helper.ApiKeyHelper.findByKey(ApiKeyHelper.java:28) ~[com.develance.vlastni.common-0.0.2-SNAPSHOT.jar:?]
        at com.develance.vlastni.api.auth.ApiKeyServletResourceHelperImpl.findByKey(ApiKeyServletResourceHelperImpl.java:50) ~[ApiKeyServletResourceHelperImpl.class:?]
        at com.develance.vlastni.api.auth.AuthServlet.doGet(AuthServlet.java:44) [AuthServlet.class:?]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:618) [servlet-api.jar:?]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:725) [servlet-api.jar:?]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:291) [catalina.jar:8.0.20]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206) [catalina.jar:8.0.20]
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52) [tomcat-websocket.jar:8.0.20]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239) [catalina.jar:8.0.20]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206) [catalina.jar:8.0.20]
        at org.apache.catalina.filters.CorsFilter.handleSimpleCORS(CorsFilter.java:303) [catalina.jar:8.0.20]
        at org.apache.catalina.filters.CorsFilter.doFilter(CorsFilter.java:167) [catalina.jar:8.0.20]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239) [catalina.jar:8.0.20]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206) [catalina.jar:8.0.20]
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:219) [catalina.jar:8.0.20]
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106) [catalina.jar:8.0.20]
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:501) [catalina.jar:8.0.20]
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:142) [catalina.jar:8.0.20]
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79) [catalina.jar:8.0.20]
        at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:610) [catalina.jar:8.0.20]
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88) [catalina.jar:8.0.20]
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:516) [catalina.jar:8.0.20]
        at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1086) [tomcat-coyote.jar:8.0.20]
        at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:659) [tomcat-coyote.jar:8.0.20]
        at org.apache.coyote.http11.Http11NioProtocol$Http11ConnectionHandler.process(Http11NioProtocol.java:223) [tomcat-coyote.jar:8.0.20]
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1558) [tomcat-coyote.jar:8.0.20]
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1515) [tomcat-coyote.jar:8.0.20]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_101]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_101]
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-util.jar:8.0.20]
        at java.lang.Thread.run(Thread.java:745) [?:1.8.0_101]
Caused by: com.orientechnologies.orient.enterprise.channel.binary.ONetworkProtocolException: Cannot read protocol version from remote server api.vlasni.com/45.56.122.213:2424: java.net.SocketTime$
        at com.orientechnologies.orient.client.binary.OChannelBinaryAsynchClient.<init>(OChannelBinaryAsynchClient.java:94) ~[orientdb-client-2.2.11.jar:2.2.11]
        at com.orientechnologies.orient.client.remote.ORemoteConnectionPool.createNetworkConnection(ORemoteConnectionPool.java:58) ~[orientdb-client-2.2.11.jar:2.2.11]
        ... 42 more
```
So I stopped applications, tried to shutdown Orient softly but it wouldn't stop, so I ended up killing it.
Below is the output from orientdb.err from that moment:

```
java.lang.OutOfMemoryError: Java heap space
        at com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OLogSegment.commitLog(OLogSegment.java:170)
        at com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OLogSegment.access$100(OLogSegment.java:33)
        at com.orientechnologies.orient.core.storage.impl.local.paginated.wal.OLogSegment$FlushTask.run(OLogSegment.java:138)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
$ANSI{green {db=vlasni3}} Error executing request
java.lang.OutOfMemoryError: Java heap space
Error on client connection
java.lang.OutOfMemoryError: Java heap space
$ANSI{green {db=vlasni3}} Error executing request
java.lang.OutOfMemoryError: Java heap space
Error during WAL background flush
java.lang.OutOfMemoryError: Java heap space
$ANSI{green {db=vlasni3}} Error on fetching record during browsing. The record has been skipped
com.orientechnologies.orient.core.exception.ODatabaseException: Error on retrieving record #15:47027 (cluster: poi)
        DB name="vlasni3"
        at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.executeReadRecord(ODatabaseDocumentTx.java:1926)
        at com.orientechnologies.orient.core.tx.OTransactionNoTx.loadRecord(OTransactionNoTx.java:107)
        at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.load(ODatabaseDocumentTx.java:1660)
        at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.load(ODatabaseDocumentTx.java:103)
        at com.orientechnologies.orient.core.iterator.OIdentifiableIterator.readCurrentRecord(OIdentifiableIterator.java:290)
        at com.orientechnologies.orient.core.iterator.ORecordIteratorClusters.hasNext(ORecordIteratorClusters.java:167)
        at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.serialIterator(OCommandExecutorSQLSelect.java:1617)
        at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.fetchFromTarget(OCommandExecutorSQLSelect.java:1566)
        at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.executeSearch(OCommandExecutorSQLSelect.java:525)
        at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.execute(OCommandExecutorSQLSelect.java:488)
        at com.orientechnologies.orient.core.sql.OCommandExecutorSQLDelegate.execute(OCommandExecutorSQLDelegate.java:74)
        at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.executeCommand(OAbstractPaginatedStorage.java:2672)
        at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.command(OAbstractPaginatedStorage.java:2618)
        at com.orientechnologies.orient.core.command.OCommandRequestTextAbstract.execute(OCommandRequestTextAbstract.java:69)
        at com.orientechnologies.orient.server.network.protocol.binary.ONetworkProtocolBinary.command(ONetworkProtocolBinary.java:1438)
        at com.orientechnologies.orient.server.network.protocol.binary.ONetworkProtocolBinary.executeRequest(ONetworkProtocolBinary.java:576)
        at com.orientechnologies.orient.server.network.protocol.binary.ONetworkProtocolBinary.sessionRequest(ONetworkProtocolBinary.java:321)
        at com.orientechnologies.orient.server.network.protocol.binary.ONetworkProtocolBinary.execute(ONetworkProtocolBinary.java:197)
        at com.orientechnologies.common.thread.OSoftThread.run(OSoftThread.java:77)
Caused by: java.lang.OutOfMemoryError: GC overhead limit exceeded
$ANSI{green {db=vlasni3}} Error on fetching record during browsing. The record has been skipped
com.orientechnologies.orient.core.exception.ODatabaseException: Error on retrieving record #15:47072 (cluster: poi)
        DB name="vlasni3"
        at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.executeReadRecord(ODatabaseDocumentTx.java:1926)
        at com.orientechnologies.orient.core.tx.OTransactionNoTx.loadRecord(OTransactionNoTx.java:107)
        at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.load(ODatabaseDocumentTx.java:1660)
        at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.load(ODatabaseDocumentTx.java:103)
        at com.orientechnologies.orient.core.iterator.OIdentifiableIterator.readCurrentRecord(OIdentifiableIterator.java:290)
        at com.orientechnologies.orient.core.iterator.ORecordIteratorClusters.hasNext(ORecordIteratorClusters.java:167)
        at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.serialIterator(OCommandExecutorSQLSelect.java:1617)
        at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.fetchFromTarget(OCommandExecutorSQLSelect.java:1566)
        at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.executeSearch(OCommandExecutorSQLSelect.java:525)
        at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.execute(OCommandExecutorSQLSelect.java:488)
        at com.orientechnologies.orient.core.sql.OCommandExecutorSQLDelegate.execute(OCommandExecutorSQLDelegate.java:74)
        at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.executeCommand(OAbstractPaginatedStorage.java:2672)
        at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.command(OAbstractPaginatedStorage.java:2618)
        at com.orientechnologies.orient.core.command.OCommandRequestTextAbstract.execute(OCommandRequestTextAbstract.java:69)
        at com.orientechnologies.orient.server.network.protocol.binary.ONetworkProtocolBinary.command(ONetworkProtocolBinary.java:1438)
        at com.orientechnologies.orient.server.network.protocol.binary.ONetworkProtocolBinary.executeRequest(ONetworkProtocolBinary.java:576)
        at com.orientechnologies.orient.server.network.protocol.binary.ONetworkProtocolBinary.sessionRequest(ONetworkProtocolBinary.java:321)
        at com.orientechnologies.orient.server.network.protocol.binary.ONetworkProtocolBinary.execute(ONetworkProtocolBinary.java:197)
        at com.orientechnologies.common.thread.OSoftThread.run(OSoftThread.java:77)
Caused by: java.lang.OutOfMemoryError: GC overhead limit exceeded
Error on client connection
java.lang.OutOfMemoryError: Java heap space
Error during WAL background flush
java.lang.OutOfMemoryError: Java heap space
Error on client connection
java.lang.OutOfMemoryError: Java heap space
Error during WAL background flush
java.lang.OutOfMemoryError: Java heap space
Error on client connection
java.lang.OutOfMemoryError: Java heap space
Error during WAL background flush
java.lang.OutOfMemoryError: Java heap space

2016-10-11 09:42:33:101 WARNI Received signal: SIGTERM [OSignalHandler]Error on client connection
java.lang.OutOfMemoryError: Java heap space
Error on client connection
java.lang.OutOfMemoryError: Java heap space
```

What should I do to get DB back working? Keep on waiting for recovery process to complete? Go again through the process of creating a new DB and copying files from current DB?
Thanks

**Comment by vlasni** [comment]
Well, I had to proceed as I did before: created a new DB, stopped Orient, copied contents from old DB directory to new DB directory, started Orient.
New DB is working OK now again.

**Comment by taburet** [comment]
Just a status update w/o a solution. Analyzed the recovery journal from the provided OSystem database, it contains broken recovery record of 1.6GB in size. So on default heap size settings it can't fit in the heap, that's why "java.lang.OutOfMemoryError: Java heap space" was produced and recovery procedure was aborted. Anyway, the record is definitely broken, it doesn't contain valid information, raising the heap size won't fix it.

Trying to understand what can be the cause of this. Did you see `wal_backup` directory in any of database directories? Did you copy recovery related files (*.wal, *.wmr) to the new directory from the old vlasni database? How much time does it take on average to shutdown a vlasni database after a heavy write load? Do you use SSD or HDD as a storage?

**Comment by taburet** [comment]
BTW, what is your JVM heap size? There are too many "java.lang.OutOfMemoryError: GC overhead limit exceeded" and "java.lang.OutOfMemoryError: Java heap space" even when allocating objects of a relatively small size. Either the heap size is too small or something is leaking or something is broken and consumes too much memory. But according to the provided hprof file there are no apparent leaks or too large allocations, maybe it was taken w/o load on the database.

**Comment by vlasni** [comment]
@taburet 
- I only saw once the `wal_backup` directory, and it was the last time I commented.
- When copying from oldDb to newDb I did copy everything, including .wal and .wmr. (first attempt I did was copying only the files clearly related to vlasniDb but it didn't work).
- I haven't measured how much time it takes to shutdown db under/after heavy load. For what I've seen so far it takes forever.
- This is a Linode box, and it claims to be SSD.
- Config:
`ORIENTDB_OPTS_MEMORY="-Xms1024m -Xmx1024m"`
`ORIENTDB_SETTINGS="-Dstorage.diskCache.bufferSize=7424"`  (this last one I missed setting it when updated the snapshot you provided, so I will be restarting db with this set)

Thanks

**Comment by vlasni** [comment]
I just shutdown db -and it was under some heavy load- and it took almost no time, immediately shutdown. This hasn't happened always, sometimes it stays there and seems to be non-responsive to the shutdown command.

**Comment by taburet** [comment]
Could you please share that `wal_backup` directory and make a heap/thread dump next time `java.lang.OutOfMemoryError` start to appear.

**Comment by vlasni** [comment]
Sure, here is the link: https://drive.google.com/file/d/0B1qygo8te6MeZ1VJUVdVX2JEQUE/view?usp=sharing 
I forgot to mention the `wal_backup` directory is in OSystem folder not in vlasniDb folder.

**Comment by vlasni** [comment]
Just an update: maybe the cause for the OOM was the lack of setting  `ORIENTDB_SETTINGS="-Dstorage.diskCache.bufferSize=7424"`?
What I am monitoring so far shows (under similar load as before):
- mem usage reduced from 78.5% to 70%
- cpu usage decreased about 30%

Still, I think we should wait a bit longer to see how it evolves.
Will keep you updated.
Thanks

**Comment by vlasni** [comment]
Update: db has been stable last 7 hours, no issues so far, no increase in mem or cpu consumption.
Log shows nothing new.

**Comment by taburet** [comment]
Hi @vlasni! I don't think the lack of `Dstorage.diskCache.bufferSize` may be the cause of this. Most probably it's due to low JVM's `-Xmx` value with the combination of some big queries. I suggest you to raise `-Xmx` more, to something like 1.5-2GB, just as a precaution. Closing the issue for now, since OOM situations are not handled by current ODB design.

**Comment by vlasni** [comment]
ok @taburet  I will raise `-Xmx` and see how it behaves.
And you're right about the big queries: there are about 3 queries that are very expensive and I am trying to improve them.
thank you

**Comment by vlasni** [comment]
hi @taburet ,
As I began to see these entries in log:
```
2016-10-13 09:09:03:236 SEVER Internal server error:
java.lang.NullPointerException [ONetworkProtocolHttpDb]
2016-10-13 09:09:09:120 SEVER Internal server error:
java.lang.NullPointerException [ONetworkProtocolHttpDb]
2016-10-13 09:16:41:442 SEVER Internal server error:
java.lang.NullPointerException [ONetworkProtocolHttpDb]
```
I stopped all applications and tried to shutdown ODB. This is the entry in log:
```
2016-10-13 10:58:40:493 WARNI Received signal: SIGTERM [OSignalHandler]
```
It's been there since more than 1 hour and I see ODB still running consuming 65.9% mem and with a server load of 4.4 (this is a 6 CPUs box).
Should I keep on waiting for ODB to shutdown properly or is it safe to kill it?

Thanks

**Comment by taburet** [comment]
Please take a thread dump before killing it. What's new in the server log? :)

**Comment by vlasni** [comment]
@taburet I just came back from a meeting and ODB had finally shutdown, so I didn't need to kill it.
Nothing else in the log, just what I sent in latest comments:
```
2016-10-13 09:09:03:236 SEVER Internal server error:
java.lang.NullPointerException [ONetworkProtocolHttpDb]
2016-10-13 09:09:09:120 SEVER Internal server error:
java.lang.NullPointerException [ONetworkProtocolHttpDb]
2016-10-13 09:16:41:442 SEVER Internal server error:
java.lang.NullPointerException [ONetworkProtocolHttpDb]
```

**Comment by laa** [comment]
@wolf4ood could you provide better logging support for HTTP client ?

**Comment by maggiolo00** [comment]
@laa 
that should be the log of the server.
By enabling debug mode it should print out the stack trace 
but i can catch special exception like java.lang.NullPointerException and print the stack trace WDYT?

**Comment by laa** [comment]
@maggiolo00 what you mean  by enabling debug mode )) ? Changing the log level ?

**Comment by maggiolo00** [comment]
@laa yes :)

**Comment by maggiolo00** [comment]
@vlasni 

which operation do you do with http protocol ?

**Comment by vlasni** [comment]
@maggiolo00 
I don't do anything expressly with http. I understand I am using it only when in Studio.

**Comment by maggiolo00** [comment]
@laa 
added extra logging in case of NPE

**Comment by laa** [comment]
@vlasni could you run your queries under the latest version of 2.2.x branch ?

**Comment by vlasni** [comment]
@laa 
If you could create that build it would be easier for me as I am in the middle of too many things.
Thanks

**Comment by laa** [comment]
@vlasni could you try latest 2.2.13 version , it should contain this change.

**Comment by vlasni** [comment]
@laa, I upgraded to 2.2.13 as soon as it was available and the issue didn't show up again, so i think this can be closed.
Thanks

**Comment by maggiolo00** [comment]
Thanks @vlasni
